File: pset6/credit.py
import cs50

def main():
    while(True):
        print("Number: ", end="")
        card= cs50.get_int()
        if card>0:
            break
        print("Please enter a valid card number")
    checksum=step1(card)
    checksum=step2(checksum,card)
    length=count(card)
    firsttwo=get_firsttwo(card, length)
    if checksum%10==0:
        
        if length==15 and firsttwo==34:
            print("AMEX")
            
        elif length==15 and firsttwo==37:
            print("AMEX")
            
        elif length==16 and firsttwo>50 and firsttwo<56:
            print("MASTERCARD")
            
        elif length==16 and int(firsttwo/10)==4:
            print("VISA")
            
        elif length==13 and int(firsttwo/10)==4:
            print("VISA")
            
    else:
        print("INVALID")

def step1(n):
    checksum=0
    while(n>9):
        digit=int((n/10)%10)
        if digit*2>=10:
            checksum+=int(digit*2%10)
            checksum+=1
        else:
            checksum+=digit*2
        n=n//100
    return checksum

def step2(checksum, n):
    k=0
    while(n>1):
        digit=int(n%10)
        k+=digit
        n=n/100
    checksum+=k
    return checksum

def count(n):
    counter=0
    while(n>10):
        counter+=1
        n=int(n/10)
    counter+=1
    return counter

def get_firsttwo(n, length):
    if length<3:
        print("INVALID")
        exit(1)
    else:
        n=int(n/(10**(length-2)))
    return n


if __name__ == "__main__":
    main()

File: pset1/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int checksum, counter;
long number;
void step1_2(int k);

void step1(long j)
{
while (j>9){
int digit=(j/10)%10;
step1_2(digit);
j=j/100;
}
}
void step1_2(int k)
{
    k=k*2;
   if(k<10)
   {
       checksum+=k;
   }
   else
   {
       checksum+=k%10;
       checksum+=1;
   }
}
void step2(long p)
{
   while (p>0)
{
int digit=p%10;
checksum+=digit;
p=p/100;
}
}

void length(long u)
{
    counter=0;
    while(u>9)
    {
    u/=10;
    counter++;
    }
    counter+=1;
}

int main(void){
    
do{
printf("Number: ");
number=get_long_long();
}
while (number<0);

step1(number);
step2(number);
length(number);
int first_two=number/pow(10,counter-2);
if (checksum%10==0)
{
    if (counter==15 && first_two==34)
    {
        printf("AMEX\n");
    }
    else if (counter==15 && first_two==37)
    {
        printf("AMEX\n");
    }
    else if (counter==16 && 50<first_two && first_two<56)
    {
        printf("MASTERCARD\n");
    }
    else if (counter==16 && first_two/10==4)
    {
        printf("VISA\n");
    }
    else if(counter==13 && first_two/10==4)
    {
        printf("VISA\n");
    }
}
else {
    printf("INVALID\n");
}
}